usecases/memwatch/monitor.go

```diff
@@ -75,25 +93,162 @@ func (m *Monitor) CheckAlloc(sizeInBytes int64) error {
 	m.mu.Lock()
 	defer m.mu.Unlock()
 
-	if float64(m.used+sizeInBytes)/float64(m.limit) > m.maxRatio {
+	if float64(m.usedMemory+sizeInBytes)/float64(m.limit) > m.maxRatio {
 		return ErrNotEnoughMemory
 	}
 
 	return nil
 }
 
+func (m *Monitor) CheckMappingAndReserve(numberMappings int64, reservationTimeInS int) error {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	if reservationTimeInS > len(m.reservedMappingsBuffer) {
+		reservationTimeInS = len(m.reservedMappingsBuffer)
+	}
+
+	// expire old mappings
+	now := time.Now()
+	m.reservedMappings -= clearReservedMappings(m.lastReservationsClear, now, m.reservedMappingsBuffer)
+
+	if m.usedMappings+numberMappings+m.reservedMappings > m.maxMemoryMappings {
+		return ErrNotEnoughMappings
+	}
+	if reservationTimeInS > 0 {
+		m.reservedMappings += numberMappings
+		m.reservedMappingsBuffer[(now.Second()+reservationTimeInS)%60] += numberMappings
+	}
+
+	m.lastReservationsClear = now
+
+	return nil
+}
+
+func clearReservedMappings(lastClear time.Time, now time.Time, reservedMappingsBuffer []int64) int64 {
+	clearedMappings := int64(0)
+	if now.Sub(lastClear) >= time.Minute {
+		for i := 0; i < len(reservedMappingsBuffer); i++ {
+			clearedMappings += reservedMappingsBuffer[i]
+			reservedMappingsBuffer[i] = 0
+		}
+	} else if now.Second() == lastClear.Second() {
+		// do nothing
+	} else if now.Second() > lastClear.Second() {
+		// the value of the last refresh was already cleared
+		for i := lastClear.Second() + 1; i <= now.Second(); i++ {
+			clearedMappings += reservedMappingsBuffer[i]
+			reservedMappingsBuffer[i] = 0
+		}
+	} else {
+		// wrap around, the value of the last refresh was already cleared
+		for i := lastClear.Second() + 1; i < 60; i++ {
+			clearedMappings += reservedMappingsBuffer[i]
+			reservedMappingsBuffer[i] = 0
+		}
+		for i := 0; i < now.Second(); i++ {
+			clearedMappings += reservedMappingsBuffer[i]
+			reservedMappingsBuffer[i] = 0
+		}
+	}
+	return clearedMappings
+}
+
 func (m *Monitor) Ratio() float64 {
 	m.mu.Lock()
 	defer m.mu.Unlock()
 
-	return float64(m.used) / float64(m.limit)
+	return float64(m.usedMemory) / float64(m.limit)
 }
 
 // obtainCurrentUsage obtains the most recent live heap from runtime/metrics
 func (m *Monitor) obtainCurrentUsage() {
 	m.setUsed(m.metricsReader())
 }
 
+func (m *Monitor) obtainCurrentMappings() {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	m.usedMappings = getCurrentMappings()
+}
+
+func getCurrentMappings() int64 {
+	switch runtime.GOOS {
+	case "linux":
+		return currentMappingsCommand("cat", "/proc/%s/maps")
```

Comment:
I think that you need to be root in order to run this command, this command will run fine in K8s bc there we run Weaviate process as root, but this might be problematic if we will be running Weaviate process as a regular user. Just wanted to make a note to be aware of it

Reply:
you need t be root for other processes - for your own processes it works without root

Reply:
I tried this on a linux VM :)